### ISSUE TITLE
CMS [AXE-CORE Medium]: Heading levels SHOULD increase by one – "On this page" component header issue

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#6338]
#  [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

**Feedback framework**

- **❗️ Must** for if the feedback must be applied 
- **⚠️Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Description

Heading levels **should** increase by one. The "On this page" component header is currently an `<h3>` but there isn't an `<h2>` between it and the `<h1>`. It is disorienting and confusing for screen reader users when heading levels are inconsistent.

### Why it matters
The underlying purpose of headers is to convey the structure of the page. For sighted users, the same purpose is achieved using different sizes of text. Text size, however, is not helpful for users of screen readers, because a screen reader identifies a header only if it is properly marked-up. When heading elements are applied correctly, the page becomes much easier to navigate for screen reader users and sighted users alike.

In the same way that sighted users can glance at a page and get a sense of its contents, users of screen readers can do the same by navigating through headings. Well written and properly ordered headings can save users, especially those who use screen readers, a lot of time and frustration.

The purpose of headings is to describe the structure of the webpage, not just highlight important text. They should be brief, clear, unique, and marked with h1 through h6 elements applied in hierarchical order. All of these qualities make headings valuable tools for screen reader users. Similar to the way sighted users can glance at a page and get a sense of its contents, screen reader users can navigate through headings. Well written and properly ordered headings can save screen reader time and frustration.

In addition to making the page more accessible, headings have other benefits since search engines use headings when filtering, ordering, and displaying results. Improving the accessibility of your site can also have the effect of making your page more findable.

## Point of Contact

**VFS Point of Contact:** Jennifer

## Acceptance Criteria

As a screen reader user, I want to navigate the hierarchy of the page content using heading levels to save time and frustration.

## Environment

* Operating System: all
* Browser: all
* Screenreading device: any
* Server destination: production

## Steps to Recreate

1. Enter `https://www.va.gov/disability/` in browser
2. Start screenreading device listed in Environment
3. Verify that the headings read skip the `<h2>` level between the `<h1>` and On this page heading

## Possible Fixes 
[Related: https://github.com/department-of-veterans-affairs/va.gov-team/issues/5125]

VSP and VSA feel this component should be a definition list. 

**If this work is happening in the CMS:** this issue should be ~~reassigned to the CMS / sitewide team~~ completed by the team that owns the template.

```html
<dl>
  <dt>On this page</dt>
  <dd><a href="#jump1">Jump 1</a></dd>
  <dd><a href="#jump2">Jump 2</a></dd>
 <dd><a href="#jump3">Jump 3</a></dd>
</dl>
```
(Trevor also found a decent way to check for IE and add/remove the tabindex that would need lifecycle methods, hooks, or custom event listeners depending on the context. )
```javascript
//  https://stackoverflow.com/a/49549464
const isIE = !!document.documentMode;

export const setTabIndex = el => {
  const target = document.getElementById(el);

  if (isIE) {
    target.setAttribute("tabindex", "-1");
  }
};

export const unsetTabIndex = el => {
  const target = document.getElementById(el);

  if (isIE) {
    target.removeAttribute("tabindex");
  }
};
```

## WCAG or Vendor Guidance 

* [Heading levels should only increase by one](https://dequeuniversity.com/rules/axe/3.4/heading-order?application=AxeChrome)

**Comment by AkucherGCIO** [comment]
03/11/20 Sprint 17 Planning
Change <H3> to <H2>

**Comment by ncksllvn** [comment]
> Trevor also found a decent way to check for IE and add/remove the tabindex that would need lifecycle methods, hooks, or custom event listeners depending on the context. 

I'm confused about this. How does `tabindex` come into play for a heading mismatch?

**Comment by jenstrickland** [comment]
@ncksllvn I'm not sure what @1Copenut meant. I took that from this ticket he and I have on the On this page component: https://github.com/department-of-veterans-affairs/va.gov-team/issues/5125. Trevor, would you speak to this, please? Grazi!

**Comment by ncksllvn** [comment]
Ah okay, it was probably included just by mistake. The fix here is merged though! Will go live today.

**Comment by jenstrickland** [comment]
@ncksllvn please ping me on Slack when it's live and ready to check? Once I validate, I'll comment and close this ticket.

**Comment by jenstrickland** [comment]
Fix validated. Thank you! (Also removed the ❗️ Launchblocker label, since 508-defect-2 was downgraded on 3/19/20, after this ticket was initially written.)
